### PR TITLE
feat: wip traffic detection

### PR DESCRIPTION
**Pull request by cdaringe**
# Description

## Contributions

- I changed x because y
- I added the new feature x.

## Checklist

- [ ] I added one or multiple labels which best describes this PR.
- [ ] I have tested the changes locally.
- [ ] This PR has a reviewer on it.
- [ ] I have validated my changes in a docker container and on Ubuntu. (Only needed for Odin or Docker Changes)

### CODE REVIEW

File: Dockerfile.odin [hunk 1]
```diff
@@ -14,7 +14,7 @@ RUN cargo chef prepare --recipe-path recipe.json
 # ------------------ #
 FROM lukemathwalker/cargo-chef:latest-rust-${RUST_VERSION} as cacher
 WORKDIR /data/odin
-RUN apt-get update && apt-get install -y cmake
+RUN apt-get update && apt-get install -y cmake libpcap-dev
```
Comment (cdaringe) [inline Dockerfile.odin:17]: pcap used to sniff for packets

File: Makefile [hunk 2]
```diff
@@ -21,7 +21,7 @@ member_clippy:
 	cargo clippy
 
 docker-build: setup
-	docker compose -f ./docker-compose.dev.yml build
+	docker compose -f ./docker-compose.dev.yml build --progress=plain
```
Comment (cdaringe) [inline Makefile:24]: because docker stdio hides critical build failure output by default

File: docker-compose.yml [hunk 3]
```diff
@@ -2,6 +2,7 @@ version: "3"
 services:
   odin:
     image: mbround18/odin:latest
+    platform: "linux/amd64"
```
Comment (cdaringe) [inline docker-compose.yml:5]: i'm on mac m1 with arm.

File: src/odin/commands/configure.rs [hunk 4]
```diff
@@ -58,6 +58,9 @@ pub struct Configuration {
 
   /// Sets the save interval in seconds
   pub save_interval: Option<u16>,
+
+  /// Steam home dirname
+  pub steam_home_dirname: String,
```
Comment (cdaringe) [inline src/odin/commands/configure.rs:63]: the repo is very much coded to only be executing in cointainer. i just _couldnt_ get the container to ever boot. the steam script segfaulting and not emitting any output :( 

 i figured "maybe i can at least unit/integration test my feature" without actually running a real server. however, 'odin configure' and 'odin install' were some pre-reqs that i only _started_ to hack through to test this approach, but well exceeded my timebox today

File: src/odin/commands/configure.rs [hunk 5]
```diff
@@ -118,12 +123,12 @@ impl Configuration {
 
   fn perform_preflight_checks(&self) -> Result<(), Box<dyn std::error::Error>> {
     let paths_to_check = vec![
-      Path::new("/home/steam/valheim"),
-      Path::new("/home/steam/scripts"),
-      Path::new("/home/steam/.bashrc"),
+      PathBuf::from_iter([&self.steam_home_dirname, "valheim"]),
+      PathBuf::from_iter([&self.steam_home_dirname, "scripts"]),
+      PathBuf::from_iter([&self.steam_home_dirname, ".bashrc"])
```
Comment (cdaringe) [inline src/odin/commands/configure.rs:128]: my attempts to support running external of the container, by de-hardcoding these values

File: src/odin/commands/start.rs [hunk 6]
```diff
@@ -1,25 +1,43 @@
...
-pub fn invoke(dry_run: bool) {
+fn run_server(config: &config::ValheimArguments) -> Child {
+  server::start_daemonized(config)
+    .map_err(|err| format!("{}", err))
+    .and_then(|it| it.map_err(|err| format!("{}", err)))
+    .map_err(|err: String| {
+      error!(target: "commands_start", "Error: {}", err);
+      exit(1);
+    })
+    .expect("failed to daemonize server")
```
Comment (cdaringe) [inline src/odin/commands/start.rs:18]: i didn't see a repo error enum, but didn't want to a) introduce one or b) go `.unwrap()` crazy, so i opted to just use `String` as the `Err` variant to get sane messaging propagated up. not married to it (or any of this really) at all!

File: src/odin/commands/start.rs [hunk 7]
```diff
@@ -1,25 +1,43 @@
...
   if !dry_run {
-    match server::start_daemonized(config) {
-      Ok(_) => info!(target: "commands_start", "Success, daemonized"),
-      Err(e) => {
-        error!(target: "commands_start", "Error: {}", e);
-        exit(1);
+    match pause_on_idle_s > 0 {
+      true => loop {
+        let mut child = run_server(&config);
+        let _ = server::spawn_kill_on_idle(&mut child, pause_on_idle_s).map_err(|err| {
```
Comment (cdaringe) [inline src/odin/commands/start.rs:32]:
```suggestion
        let _ = server::kill_on_idle(&mut child, pause_on_idle_s).map_err(|err| {
```

File: docker-compose.yml (hunk 3 above)
Comment (mbround18) [reply docker-compose.yml:5]: This should only really be in your docker-compose.dev.yaml but have you had success running valheim on m1? weve had lots of issues before with people struggling with it.

File: src/odin/server/traffic.rs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+use std::process::Child;
+
+use pcap::{Capture, Device};
+
+pub fn spawn_kill_on_idle(child: &mut Child, pause_on_idle_s: u32) -> Result<(), String> {
```
Comment (cdaringe) [inline src/odin/server/traffic.rs:5]: well, the plan was to _pause_ the process. i realized... i'd need _another_ cgroup to _pause_ anything, and i'm not _yet_ talented enough to do that. so, my first pass was going to be naively killing the pid, then rebooting it on traffic. then on a second pass, try the whole pause/resume bit, once i can learn about firing up that server in a pausable cgroup.

Comment (cdaringe) [inline src/odin/server/traffic.rs:5]:
```suggestion
pub fn kill_on_idle(child: &mut Child, pause_on_idle_s: u32) -> Result<(), String> {
```

File: src/odin/server/traffic.rs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    // @todo uh errr uhhh get the actual server port into here
+    .filter(&format!("udp port {}", "2456"), true)
+    .map_err(|it| it.to_string())?;
+
+  while let Ok(_packet) = cap.next_packet() {
+    // pass - packet was received, so we are not idle
+  }
+  child.kill().map_err(|e| e.to_string())?;
+  Ok(())
+}
```
Comment (cdaringe) [inline src/odin/server/traffic.rs:26]: really, this is probably _wrong_ to block the main thread. should probably have this as a bg thread, with an `Arc<Mutex<Child>>` that is swapped on restart

File: docker-compose.yml (hunk 3 above)
Comment (cdaringe) [reply docker-compose.yml:5]: arrggg no i couldn't dev at all! i should have searched earlier 😢

File: src/odin/server/traffic.rs (hunk 8 above)
Comment (cdaringe) [reply src/odin/server/traffic.rs:5]: arrrg i'm such a dummy! the proc does respond to SIGSTOP/SIGCONT already. no cgroups required

**Comment by mbround18** [comment]
@cdaringe this is a really good first pass! :) ill start diving through it later tonight, apologies on delay man ive been busy remodeling a kitchen

**Comment by cdaringe** [comment]
closing for now, just to halt the constant auto-merges 😄

**Pull request closed without merging**